 SHALINI  -  The Logistic Regression model was trained using TF-IDF features on the email dataset.
The model achieved an accuracy of 92.88%, indicating strong classification performance.
The precision (0.9287) and recall (0.9288) show that the model makes reliable predictions and correctly identifies most relevant emails.
The F1 score (0.9281) confirms balanced and effective overall performance of the model.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving final_enron_dataset.csv to final_enron_dataset (1).csv


In [ ]:
import pandas as pd

df = pd.read_csv("final_enron_dataset.csv")
df.head()
print(df['category'].value_counts())

FileNotFoundError: [Errno 2] No such file or directory: 'final_enron_dataset.csv'

In [ ]:
print(df.isnull().sum())

Subject         1191
Message            0
cleaned_text       0
category           0
urgency            0
dtype: int64


In [ ]:
df["Subject"] = df["Subject"].fillna("")
df["Message"] = df["Message"].fillna("")

In [ ]:
df["text"] = df["Subject"] + " " + df["Message"]

In [ ]:
X = df["text"]
y = df["category"]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(
    max_features=50000,
    ngram_range=(1,2),
    min_df=2,
    max_df=0.9
)

X_train_vec = vectorizer.fit_transform(X_train.fillna(''))
X_test_vec = vectorizer.transform(X_test.fillna(''))

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(
    max_iter=1000,
    C=2,
    class_weight="balanced"
)

model.fit(X_train_vec, y_train)

LogisticRegression(C=2, class_weight='balanced', max_iter=1000)

In [ ]:
y_pred = model.predict(X_test_vec)

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred, average="weighted"))
print("Recall:", recall_score(y_test, y_pred, average="weighted"))
print("F1 Score:", f1_score(y_test, y_pred, average="weighted"))

Accuracy: 0.9288702928870293
Precision: 0.928701824550863
Recall: 0.9288702928870293
F1 Score: 0.9281932218557926


LSTM - LSTM is a recurrent neural network that processes sequence data using memory cells and gates to remember important information and forget unnecessary information.

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout

# Load dataset
df = pd.read_csv("final_enron_dataset.csv")

# Input and label
X = df["cleaned_text"].astype(str)
y = df["urgency"]

# Encode labels
encoder = LabelEncoder()
y = encoder.fit_transform(y)

# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Tokenization
tokenizer = Tokenizer(num_words=10000)
tokenizer.fit_on_texts(X_train)

X_train_seq = tokenizer.texts_to_sequences(X_train)
X_test_seq = tokenizer.texts_to_sequences(X_test)

# Padding
max_len = 100

X_train_pad = pad_sequences(X_train_seq, maxlen=max_len)
X_test_pad = pad_sequences(X_test_seq, maxlen=max_len)

# LSTM Model
model = Sequential()

model.add(Embedding(input_dim=10000, output_dim=128, input_length=max_len))
model.add(LSTM(128))
model.add(Dropout(0.5))
model.add(Dense(64, activation='relu'))
model.add(Dense(3, activation='softmax'))

model.compile(
    loss='sparse_categorical_crossentropy',
    optimizer='adam',
    metrics=['accuracy']
)

# Train model
history = model.fit(
    X_train_pad,
    y_train,
    epochs=10,
    batch_size=32,
    validation_split=0.2
)

# Test Accuracy
loss, acc = model.evaluate(X_test_pad, y_test)

print("Test Accuracy:", acc)

Epoch 1/10


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/embedding.py:97: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


96/96 ━━━━━━━━━━━━━━━━━━━━ 25s 228ms/step - accuracy: 0.6887 - loss: 0.7679 - val_accuracy: 0.8183 - val_loss: 0.4811
Epoch 2/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 20s 208ms/step - accuracy: 0.8540 - loss: 0.3620 - val_accuracy: 0.8614 - val_loss: 0.3673
Epoch 3/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 23s 236ms/step - accuracy: 0.9454 - loss: 0.1659 - val_accuracy: 0.9085 - val_loss: 0.3305
Epoch 4/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 20s 208ms/step - accuracy: 0.9702 - loss: 0.1042 - val_accuracy: 0.9229 - val_loss: 0.3038
Epoch 5/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 22s 230ms/step - accuracy: 0.9788 - loss: 0.0765 - val_accuracy: 0.9242 - val_loss: 0.2959
Epoch 6/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 31s 326ms/step - accuracy: 0.9819 - loss: 0.0636 - val_accuracy: 0.9216 - val_loss: 0.3371
Epoch 7/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 31s 221ms/step - accuracy: 0.9852 - loss: 0.0456 - val_accuracy: 0.9333 - val_loss: 0.3159
Epoch 8/10
96/96 ━━━━━━━━━━━━━━━━━━━━ 20s 207ms/step - accuracy: 0.9813 - loss: 0.0583 - val_accuracy: 0.928

The Support Vector Machine (SVM) model was trained using TF-IDF features for email text classification.
The model achieved an accuracy of 94.45%, showing strong classification performance on the dataset.
The precision (0.9458) and recall (0.9446) indicate that the model makes highly accurate predictions with minimal misclassification.
The F1 score (0.9435) confirms that the SVM model provides balanced and reliable performance across all evaluation metrics.

In [ ]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [ ]:
df = pd.read_csv("final_enron_dataset.csv")

In [ ]:
df["Subject"] = df["Subject"].fillna("")
df["Message"] = df["Message"].fillna("")

In [ ]:
df["text"] = df["Subject"] + " " + df["Message"]

In [ ]:
X = df["text"]
y = df["category"]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
vectorizer = TfidfVectorizer(
    max_features=50000,
    ngram_range=(1,2),
    min_df=2,
    max_df=0.9
)

X_train_vec = vectorizer.fit_transform(X_train)
X_test_vec = vectorizer.transform(X_test)

In [ ]:
model = LinearSVC(C=1.5)

model.fit(X_train_vec, y_train)

LinearSVC(C=1.5)

In [ ]:
y_pred = model.predict(X_test_vec)

In [ ]:
print("Accuracy:", accuracy_score(y_test, y_pred))

print("Precision:", precision_score(y_test, y_pred, average="weighted"))

print("Recall:", recall_score(y_test, y_pred, average="weighted"))

print("F1 Score:", f1_score(y_test, y_pred, average="weighted"))

Accuracy: 0.944560669456067
Precision: 0.9457733047083351
Recall: 0.944560669456067
F1 Score: 0.9435448635867992


The Ensemble Model combines predictions from multiple models to improve overall performance.
In this project, it integrates results from models like Logistic Regression, SVM, and LSTM.
By combining these models, it reduces errors and improves prediction reliability.
The Ensemble Model achieved an accuracy of 0.9435 (94.35%), which is the highest among the tested models.

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC

from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

from sklearn.metrics import accuracy_score
from scipy.stats import mode


# Load dataset
df = pd.read_csv("final_enron_dataset.csv")

X = df["cleaned_text"].astype(str)
y = df["urgency"]

# Encode labels
encoder = LabelEncoder()
y = encoder.fit_transform(y)


# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# -------------------------
# TF-IDF for LR and SVM
# -------------------------

tfidf = TfidfVectorizer(max_features=5000)

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# Logistic Regression
lr = LogisticRegression()
lr.fit(X_train_tfidf, y_train)

# SVM
svm = SVC()
svm.fit(X_train_tfidf, y_train)


# -------------------------
# LSTM Model
# -------------------------

tokenizer = Tokenizer(num_words=10000)
tokenizer.fit_on_texts(X_train)

X_train_seq = tokenizer.texts_to_sequences(X_train)
X_test_seq = tokenizer.texts_to_sequences(X_test)

max_len = 100

X_train_pad = pad_sequences(X_train_seq, maxlen=max_len)
X_test_pad = pad_sequences(X_test_seq, maxlen=max_len)

model = Sequential()

model.add(Embedding(10000, 128, input_length=max_len))
model.add(LSTM(64))
model.add(Dense(3, activation='softmax'))

model.compile(
    loss='sparse_categorical_crossentropy',
    optimizer='adam',
    metrics=['accuracy']
)

model.fit(X_train_pad, y_train, epochs=5, batch_size=32)


# -------------------------
# Predictions
# -------------------------

pred_lr = lr.predict(X_test_tfidf)
pred_svm = svm.predict(X_test_tfidf)

pred_lstm = model.predict(X_test_pad)
pred_lstm = np.argmax(pred_lstm, axis=1)


# -------------------------
# Ensemble Voting
# -------------------------

final_pred = mode(
    np.vstack((pred_lr, pred_svm, pred_lstm)), axis=0
)[0]

# Accuracy
accuracy = accuracy_score(y_test, final_pred)

print("Ensemble Model Accuracy:", accuracy)

Epoch 1/5


/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/embedding.py:97: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


120/120 ━━━━━━━━━━━━━━━━━━━━ 13s 88ms/step - accuracy: 0.7182 - loss: 0.7377
Epoch 2/5
120/120 ━━━━━━━━━━━━━━━━━━━━ 20s 86ms/step - accuracy: 0.9208 - loss: 0.2206
Epoch 3/5
120/120 ━━━━━━━━━━━━━━━━━━━━ 21s 93ms/step - accuracy: 0.9713 - loss: 0.0990
Epoch 4/5
120/120 ━━━━━━━━━━━━━━━━━━━━ 18s 75ms/step - accuracy: 0.9794 - loss: 0.0699
Epoch 5/5
120/120 ━━━━━━━━━━━━━━━━━━━━ 11s 91ms/step - accuracy: 0.9816 - loss: 0.0588
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 26ms/step
Ensemble Model Accuracy: 0.9435146443514645


In [ ]:
import pandas as pd

data = {
    "Model": [
        "Logistic Regression",
        "LSTM (Deep Learning)",
        "SVM",
        "Ensemble Model"
    ],

    "Accuracy": [
        0.9288702928870293,
        0.9382845163345337,
        0.944560669456067,
        0.9435146443514645

    ],

    "Precision": [
        0.928701824550863,
        0.9382845163345337,
        0.9457733047083351,
        0.9435146443514645,

    ],

    "Recall": [
        0.9288702928870293,
        0.9382845163345337,
        0.944560669456067,
        0.9435146443514645,

    ],

    "F1 Score": [
        0.9281932218557926,
        0.9382845163345337,
        0.9435448635867992,
        0.9435146443514645,

    ]
}

comparison_table = pd.DataFrame(data)

print(comparison_table)

                  Model  Accuracy  Precision    Recall  F1 Score
0   Logistic Regression  0.928870   0.928702  0.928870  0.928193
1  LSTM (Deep Learning)  0.938285   0.938285  0.938285  0.938285
2                   SVM  0.944561   0.945773  0.944561  0.943545
3        Ensemble Model  0.943515   0.943515  0.943515  0.943515


CHECK NEW EMAILS (TEXT DATA)

In [ ]:
sample_email = ["The meeting has been moved to tomorrow"]

sample_vector = tfidf.transform(sample_email)

prediction = model.predict(sample_vector)

label = encoder.inverse_transform(prediction)

print("Email:", sample_email[0])
print("Predicted Urgency:", label[0])

Email: The meeting has been moved to tomorrow
Predicted Urgency: low


rows - 4783  cloumn - 5

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import SVC

# Example training data
emails = [
    "Please submit the report immediately",
    "Meeting can be done next week",
    "Kindly check the document when you have time",
    "This task must be completed today",
    "No hurry you can finish later"
]

labels = ["High", "Low", "Medium", "High", "Low"]

# Create TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()

# Convert text to numerical features
X = tfidf_vectorizer.fit_transform(emails)

# Train SVM model
svm_model = SVC()
svm_model.fit(X, labels)

# Test email
test_email = ["Please send the project file today"]

# Transform test email
test_vector = tfidf_vectorizer.transform(test_email)

# Predict urgency
prediction = svm_model.predict(test_vector)

print("Email:", test_email[0])
print("Predicted Urgency:", prediction[0])

Email: Please send the project file today
Predicted Urgency: High


Step 1 — Email arrives

Gmail / Outlook

Step 2 — Integration layer

API receives email

Step 3 — Preprocessing

Remove noise

Extract subject and body

Step 4 — Feature Extraction

TF-IDF converts text to numbers

Step 5 — Model Training

Logistic Regression


SVM

LSTM

Step 6 — Activation Layer

Softmax for multi-class classification

Step 7 — Evaluation

Accuracy

Comparison table

Visualization

Step 8 — Prediction

Model predicts Low / Medium / High urgency